pkg/apiserver/handlers/networkpolicyanalysis/handler.go

```diff
@@ -0,0 +1,249 @@
+// of the policy UIDs, and manually inserts Kubernetes NetworkPolicy default isolation
+// rules if exists. The default isolation rule's direction depends on ifSource,
+// and has the index of -1 to indicate lower precedence.
+func processQueryResults(endpoints *networkpolicy.EndpointNetworkPolicyRules, ifSource bool) (sets.Set[types.UID], []*networkpolicy.RuleInfo) {
+	policyUIDs := sets.New[types.UID]()
+	isolation := make([]*networkpolicy.RuleInfo, 0)
+	for _, internalPolicy := range endpoints.AppliedPolicies {
+		policyUIDs.Insert(internalPolicy.SourceRef.UID)
+		if internalPolicy.SourceRef.Type == controlplane.K8sNetworkPolicy {
+			// check if the Kubernetes NetworkPolicy creates ingress or egress isolation
+			for _, rule := range internalPolicy.Rules {
+				if rule.Direction == controlplane.DirectionIn && !ifSource {
+					isolation = append(isolation, &networkpolicy.RuleInfo{Policy: internalPolicy, Index: -1, Direction: v1beta2.DirectionIn})
+				} else if rule.Direction == controlplane.DirectionOut && ifSource {
+					isolation = append(isolation, &networkpolicy.RuleInfo{Policy: internalPolicy, Index: -1, Direction: v1beta2.DirectionOut})
+				}
+			}
+		}
+	}
+	return policyUIDs, isolation
+}
+
+// predictEndpointsRules returns the predicted rules effective from srcEndpoints to dstEndpoints.
+// Rules returned satisfy a. in source applied policies and destination egress rules,
+// or b. in source ingress rules and destination applied policies or c. applied to KNP default isolation.
+func predictEndpointsRules(srcEndpoints, dstEndpoints *networkpolicy.EndpointNetworkPolicyRules) *endpoint.Rule {
+	commonRules := make([]*networkpolicy.RuleInfo, 0)
+	if srcEndpoints != nil && dstEndpoints != nil {
+		srcPolicies, srcIsolated := processQueryResults(srcEndpoints, true)
+		dstPolicies, dstIsolated := processQueryResults(dstEndpoints, false)
+		for _, rule := range dstEndpoints.EgressRules {
+			if srcPolicies.Has(rule.Policy.SourceRef.UID) {
+				commonRules = append(commonRules, rule)
+			}
+		}
+		for _, rule := range srcEndpoints.IngressRules {
+			if dstPolicies.Has(rule.Policy.SourceRef.UID) {
+				commonRules = append(commonRules, rule)
+			}
+		}
+		for _, defaultDropRule := range srcIsolated {
+			commonRules = append(commonRules, defaultDropRule)
+		}
+		for _, defaultDropRule := range dstIsolated {
+			commonRules = append(commonRules, defaultDropRule)
+		}
+	}
+	// if no common rule or default isolation is found, return empty result
+	if len(commonRules) == 0 {
+		return &endpoint.Rule{}
+	}
+	// sort the common rules based on multiple closures, the top rule has the highest precedence
+	passRule := func(r1, r2 *networkpolicy.RuleInfo) int {
+		if r2.Policy.SourceRef.Type != controlplane.K8sNetworkPolicy && r2.Policy.Rules[r2.Index].Action != nil && *r2.Policy.Rules[r2.Index].Action == v1beta1.RuleActionPass {
+			return 1
+		}
+		return 0
+	}
+	tierPriority := func(r1, r2 *networkpolicy.RuleInfo) int {
```

Comment:
Why not just return `bool` for all `lessFunc`? I thought tie breakers can be handled by subsequent lessFuncs down the lane?

Reply:
It's actually modified from https://pkg.go.dev/sort#example-package-SortMultiKeys where instead of comparing `less(p, q)` and `less(q, p)`, returning int can reduce comparisons.